cmd/ocagent/config.go

```diff
@@ -183,12 +183,19 @@ func (c *config) checkLogicalConflicts(blob []byte) error {
 			zExporterHostPort, zReceiverHostPort)
 	}
 	zExpHost, zExpPort, _ := net.SplitHostPort(zExporterHostPort)
-	zReceiverHost, zReceiverPort, _ := net.SplitHostPort(zExporterHostPort)
+	zReceiverHost, zReceiverPort, _ := net.SplitHostPort(zReceiverHostPort)
 	if eqHosts(zExpHost, zReceiverHost) && zExpPort == zReceiverPort {
 		return fmt.Errorf("ZipkinExporter address (%q) aka (%s on port %s)\nis the same as the receiver address (%q) aka (%s on port %s)",
 			zExporterHostPort, zExpHost, zExpPort, zReceiverHostPort, zReceiverHost, zReceiverPort)
 	}
 
+	// If exporter and receiver hosts are both "localhost" assume that this is some test
+	// and allow them to proceed, since it was already checked that they aren't on the
+	// same port.
+	if zExpHost == zReceiverHost && zExpHost == "localhost" {
```

Comment:
If both are "localhost" and are on the same port it is really a feedback loop, however, if you receive on port x and send to port y, as long as x != y it is a fine configuration. In manual tests I had the receiver on localhost:9410 and zipkin-docker/exporter on localhost:9411. The test above this one takes care of the equal ports, the test below if different host names maps to the same ip, but as it is they prevent the single box testing as I did.

Reply:
I unfortunately don't follow your response. 

>If both are "localhost" and are on the same port it is really a feedback loop, however, if you receive on port x and send to port y, as long as x != y it is a fine configuration. 

Right, and that's the point of what we already have in
https://github.com/census-instrumentation/opencensus-service/blob/5bd378023295ee43a9c6c0812a857d9022b7b7a6/cmd/ocagent/config.go#L178-L198

>In manual tests I had the receiver on localhost:9410 and zipkin-docker/exporter on localhost:9411. The test above this one takes care of the equal ports, the test below if different host names maps to the same ip, but as it is they prevent the single box testing as I did.

I don't follow that, from day 1, I've been doing "single box" testing where we have a receiver on one port and the exporter on another. I don't think we should add a hack for a feedback loop. If you encounter such a problem please file an isolated and reproducible bug and we can take a look at it.

Reply:
This is not a feedback loop: receiver is in one port exporter in other. I bet that our steps are a bit different: I start the Zipkin docker with default config and then launched the agent with following config and steps:

```shell
Paulos-MBP:opencensus-service pj$ cat ./cmd/ocagent/config.yaml
receivers:
    opencensus:
        address: "127.0.0.1:55678"

    zipkin:
        address: "localhost:9410"

exporters:
    zipkin:
        endpoint: "http://localhost:9411/api/v2/spans"

zpages:
    port: 55679
Paulos-MBP:opencensus-service pj$ ./bin/ocagent_darwin -c ./cmd/ocagent/config.yaml
2018/11/26 07:29:01 Configuration logical error: ZipkinExporter address ("localhost:9411") aka (localhost on port 9411)
is the same as the receiver address ("localhost:9410") aka (localhost on port 9411)
```

The one above is the small mistake about using `zExporterHostPort` instead of `zReceiverHostPort`, i.e.:
```go
	zReceiverHost, zReceiverPort, _ := net.SplitHostPort(zExporterHostPort)
```

Fixing that I hit the next one that made me add the exception to localhost:
```shell
Paulos-MBP:opencensus-service pj$ ./bin/ocagent_darwin -c ./cmd/ocagent/config.yaml
2018/11/26 07:30:51 Configuration logical error: ZipkinExporter address ("localhost:9411") aka (127.0.0.1)
is the same as the
receiver address ("localhost:9410") aka (127.0.0.1)
```

Reply:
@pjanotti that code masks the problem. The correct fix is to check also if the ports are the same so basically
```diff
diff --git a/cmd/ocagent/config.go b/cmd/ocagent/config.go
index 6a814a5..c298f0c 100644
--- a/cmd/ocagent/config.go
+++ b/cmd/ocagent/config.go
@@ -192,7 +192,7 @@ func (c *config) checkLogicalConflicts(blob []byte) error {
 	// Otherwise, now let's resolve the IPs and ensure that they aren't the same
 	zExpIPAddr, _ := net.ResolveIPAddr("ip", zExpHost)
 	zReceiverIPAddr, _ := net.ResolveIPAddr("ip", zReceiverHost)
-	if zExpIPAddr != nil && zReceiverIPAddr != nil && reflect.DeepEqual(zExpIPAddr, zReceiverIPAddr) {
+	if zExpIPAddr != nil && zReceiverIPAddr != nil && reflect.DeepEqual(zExpIPAddr, zReceiverIPAddr) && zExpPort == zReceiverPort {
 		return fmt.Errorf("ZipkinExporter address (%q) aka (%+v)\nis the same as the\nreceiver address (%q) aka (%+v)",
 			zExporterHostPort, zExpIPAddr, zReceiverHostPort, zReceiverIPAddr)
 	}
```
otherwise we'll be masking the problem and will let loopbacks happen.